Sources/DiagnosticsReporter.swift

```diff
@@ -46,13 +49,33 @@ public enum DiagnosticsReporter {
     /// - Parameters:
     ///   - reporters: The reporters to use. Defaults to `DefaultReporter.allReporters`. Use this parameter if you'd like to exclude certain reports.
     ///   - filters: The filters to use for the generated diagnostics. Should conform to the `DiagnosticsReportFilter` protocol.
-    public static func create(using reporters: [DiagnosticsReporting.Type] = DefaultReporter.allReporters, filters: [DiagnosticsReportFilter.Type]? = nil) -> DiagnosticsReport {
-        let reportChapters = reporters.map { reporter -> DiagnosticsChapter in
-            var chapter = reporter.report()
-            if let filters = filters, !filters.isEmpty {
-                chapter.applyingFilters(filters)
+    ///   - smartInsightsProvider: Provide any smart insights for the given `DiagnosticsChapter`.
+    public static func create(using reporters: [DiagnosticsReporting] = DefaultReporter.allReporters, filters: [DiagnosticsReportFilter.Type]? = nil, smartInsightsProvider: SmartInsightsProviding? = nil) -> DiagnosticsReport {
+        /// We should be able to parse Smart insights out of other chapters.
+        /// For example: read out errors from the log chapter and create insights out of it.
+        ///
+        /// Therefore, we are generating insights on the go and add them to the Smart Insights later.
+        var smartInsights: [SmartInsightProviding] = []
+
+        var reportChapters = reporters
+            .filter { ($0 is SmartInsightsReporter) == false }
+            .map { reporter -> DiagnosticsChapter in
+                var chapter = reporter.report()
+                if let filters = filters, !filters.isEmpty {
+                    chapter.applyingFilters(filters)
+                }
+                if let smartInsightsProvider = smartInsightsProvider, let insights = smartInsightsProvider.smartInsights(for: chapter) {
+                    smartInsights.append(contentsOf: insights)
```

Comment:
It seems like we already start with an empty array here, so I don't think it adds much to return an optional array for `smartInsights(for: chapter)`?

Reply:
There are two arrays here; The container `smartInsights` array which indeed is already created which is fine since there will be default smart insights.

> so I don't think it adds much to return an optional array for

I disagree. It's nitpicking here either way since the overhead of an empty array is super low, but the array returned by the `smartInsightsProvider` is a different array so has nothing to do with the already instantiated array.

Reply:
Works for me. :)